tests/unit/obuf-limits.tcl

```diff
@@ -18,65 +18,65 @@ start_server {tags {"obuf-limits"}} {
-        r config set client-output-buffer-limit {pubsub 0 100000 10}
-        set rd1 [redis_deferring_client]
-
-        $rd1 subscribe foo
-        set reply [$rd1 read]
-        assert {$reply eq "subscribe foo 1"}
-
-        set omem 0
-        set start_time 0
-        set time_elapsed 0
-        while 1 {
-            if {$start_time != 0} {
-                # Slow down loop when omen has reached the limit.
-                after 10
-            }
-            r publish foo [string repeat "x" 1000]
-            set clients [split [r client list] "\r\n"]
-            set c [split [lindex $clients 1] " "]
-            if {![regexp {omem=([0-9]+)} $c - omem]} break
-            if {$omem > 100000} {
-                if {$start_time == 0} {set start_time [clock seconds]}
-                set time_elapsed [expr {[clock seconds]-$start_time}]
-                if {$time_elapsed >= 5} break
-            }
+    
+    foreach {soft_limit_time wait_for_timeout} {3 yes
+                                                4 no } {
+        if $wait_for_timeout {
+            set test_name "Client output buffer soft limit is enforced if time is overreached"
+        } else {
+            set test_name "Client output buffer soft limit is not enforced too early and is enforced when no traffic"
         }
-        assert {$omem >= 100000 && $time_elapsed >= 5 && $time_elapsed <= 10}
-        $rd1 close
-    }
-
-    test {Client output buffer soft limit is enforced if time is overreached} {
-        r config set client-output-buffer-limit {pubsub 0 100000 3}
-        set rd1 [redis_deferring_client]
-
-        $rd1 subscribe foo
-        set reply [$rd1 read]
-        assert {$reply eq "subscribe foo 1"}
 
-        set omem 0
-        set start_time 0
-        set time_elapsed 0
-        while 1 {
-            if {$start_time != 0} {
-                # Slow down loop when omen has reached the limit.
-                after 10
+        test $test_name {
+            r config set client-output-buffer-limit "pubsub 0 100000 $soft_limit_time"
+            set soft_limit_time [expr $soft_limit_time*1000]
+            set rd1 [redis_deferring_client]
+
+            $rd1 client setname test_client
+            set reply [$rd1 read]
+            assert {$reply eq "OK"}
+
+            $rd1 subscribe foo
+            set reply [$rd1 read]
+            assert {$reply eq "subscribe foo 1"}
+
+            set omem 0
+            set start_time 0
+            set time_elapsed 0
+            while 1 {
+                r publish foo [string repeat "x" 1000]
+                set clients [split [r client list] "\r\n"]
+                set c [lsearch -inline $clients *name=test_client*]
+                if {$wait_for_timeout && $c == ""} {
+                    # Make sure we're disconnected when we reach the soft limit
+                    assert {$omem >= 100000 && $time_elapsed >= $soft_limit_time}
```

Comment:
Here the assertion appears, $time_elapsed needs to be recalculated, otherwise the connection may be disconnected in the course of the above code, $time_elapsed will be less than $soft_limit_time.

Reply:
```suggestion
                    assert {$omem >= 100000 && $time_elapsed < [expr $soft_limit_time*2]}
```
When $omem > 100000, $start_time is actually behind, ```$time_elapsed >= $soft_limit_time``` is not necessarily true.

Reply:
This is a specific test relevant only when `wait_for_timeout` is `yes`. If it's `yes` and we're disconnected (`c == ""`) then we verify the `time_elapsed` is larger than `soft_limit_time` to make sure we were not disconnected before reaching the soft limit timeout.

Reply:
@yoav-steinberg what if we (the tcl code) detected the `start_time` late?
i'm not sure we can afford to test that it wasn't disconnected prematurely.

Reply:
@oranagra if it detects `start_time` later then `time_elapsed` will also be calculated later. `time_elapsed` is always relative to the last updated `start_time`.

Reply:
@yoav-steinberg ```start_time``` is delayed, but detection of user exit is on time, which may result in ```time_elapsed``` being less than soft_limit_time.

Reply:
@yoav-steinberg There is a probability that the test will.
```
Client output buffer soft limit is enforced if time is overreached in tests/unit/obuf-limits.tcl
Expected 143528 >= 100000 && 2981 >= 3000 (context: type eval line 23 cmd {assert {$omem >= 100000 && $time_elapsed >= $soft_limit_time}} proc ::test)
```

Reply:
> Here the assertion appears, $time_elapsed needs to be recalculated, otherwise the connection may be disconnected in the course of the above code, $time_elapsed will be less than $soft_limit_time.

Thanks, now I understand what you mean. I'll do a fix? Does the problem recreate when running with `--loop`? How long did you run it for?

Reply:
@YaacovHazan No problem, I'll test it if you make changes, my freebsd only has 256m RAM so it's easy to reproduce these problems.

Reply:
It's about half an hour to show up, I'm just going to go do something else after the run and come back later to see if there are any problems.

Reply:
@YaacovHazan I also tested your last commit, and it still gives the same error.

Reply:
@sundb Can you paste the assertion error here?

Reply:
I re-ran it again.
````
[err]: Client output buffer soft limit is enforced if time is overreached in tests/unit/obuf-limits.tcl
Expected 164032 >= 100000 && 2876 >= 3000 (context: type eval line 28 cmd {assert {$omem >= 100000 && $time_elapsed >= $soft_limit_time}} proc ::test)
```

Reply:
Thanks, on what platform/environment does this happen?

Reply:
os: freebsd 12
mem: 256M
cpu: 1core

Reply:
@sundb This doesn't recreate on my freebsd (13). I ran it for over and hour with 50 parallel clients. Can you think of a reason for this failure on your setup? The current code makes sure to calculate `time_elapsed` **after** the disconnect, so there's no real way for the time to be less than `soft_limit_time`.

Reply:
Do you have a memory limit for the virtual machine? Problems often arise in slow freebsd environments.

Reply:
@sundb pushed another fix. Can you check?

Reply:
@yoav-steinberg Another fail.
```
Client output buffer soft limit is not enforced too early and is enforced when no traffic in tests/unit/obuf-limits.tcl
Expected [regexp {omem=([0-9]+)}  - omem] (context: type eval line 32 cmd {assert {[regexp {omem=([0-9]+)} $c - omem]}} proc ::test)
```

Reply:
It looks like for some reason `$c` was not `""` but was rather `" "` (a space). How can that be? Does this happen every run?
Can you try to debug this and find the reason/fix it?

Reply:
Ok, I try it.

Reply:
@yoav-steinberg The previous assert doesn't appear anymore, I'll keep testing to see.
```
Client output buffer soft limit is enforced if time is overreached in tests/unit/obuf-limits.tcl
Expected 5413 <= [expr 3000+2000] (context: type eval line 25 cmd {assert {$time_elapsed <= [expr $soft_limit_time+2000]}} proc ::test)
```
There is still a probability, but when I changed it to +3000, it became much more stable.

Reply:
@sundb, @oranagra this assertion is a safety check. It should really never happen. I tried to recreate the issue on a freebsd vm configured to run very slowly with only 256MB ram but still couldn't.
I suggest to merge this PR and cherry-pick to 6.2. In the meantime if @sundb can recreate this failure, It would be good to investigate what's causing it? How the timing can be so off? And perhaps submit another fix.

Reply:
update: I managed to recreate the the `Expected 5008 <= [expr 3000+2000]` issue on a very slow running freebsd vm. I did some more testing and I can see that the redis commands (publishing, client list) are exceptionally slow. Sometimes peaking at over 1sec. I also saw that sometimes the tcl `after 10` takes over 1sec, which is consistently the reason the tests fails. 
So what I suggest is to accept @sundb's suggestion and have a 3000 milli safety marign instead of 2000. On normal machines this test is efficient. On very slow machines this test doesn't really work as expected (the 'Client output buffer soft limit is not enforced too early and is enforced when no traffic' won't really check what it needs to). The alternative is to make all the timeouts here x10 longer, and then the test will be good on slow machines but this will slow down the tests too much.